src/analyses/relationAnalysis.ml

```diff
@@ -0,0 +1,458 @@
+    let apr' = f apr' e' in (* x = e; *)
+    let apr'' = AD.remove_vars apr' (List.map RV.local (VH.values v_ins |> List.of_enum)) in (* remove temporary g#in-s *)
+    apr''
+
+  let write_global ask getg sideg st g x =
+    if ThreadFlag.is_multi ask then
+      Priv.write_global ask getg sideg st g x
+    else (
+      let apr = st.RelationDomain.apr in
+      let g_var = RV.global g in
+      let x_var = RV.local x in
+      let apr' = AD.add_vars apr [g_var] in
+      let apr' = AD.assign_var apr' g_var x_var in
+      {st with RelationDomain.apr = apr'}
+    )
+
+  let assign_to_global_wrapper ask getg sideg st lv f =
+    match lv with
+    (* Lvals which are numbers, have no offset and their address wasn't taken *)
+    (* This means that variables of which multiple copies may be reachable via pointers are also also excluded (they have their address taken) *)
+    (* and no special handling for them is required (https://github.com/goblint/analyzer/pull/310) *)
+    | (Var v, NoOffset) when AD.varinfo_tracked v ->
+      if not v.vglob then
+        {st with apr = f st v}
+      else (
+        let v_out = Goblintutil.create_var @@ makeVarinfo false (v.vname ^ "#out") v.vtype in (* temporary local g#out for global g *)
+        let st = {st with apr = AD.add_vars st.apr [RV.local v_out]} in (* add temporary g#out *)
+        let st' = {st with apr = f st v_out} in (* g#out = e; *)
+        if M.tracing then M.trace "apron" "write_global %a %a\n" d_varinfo v d_varinfo v_out;
+        let st' = write_global ask getg sideg st' v v_out in (* g = g#out; *)
+        let apr'' = AD.remove_vars st'.apr [RV.local v_out] in (* remove temporary g#out *)
+        {st' with apr = apr''}
+      )
+    (* Ignoring all other assigns *)
+    | _ ->
+      st
+
+  let no_overflow ctx exp =
+    let ik = Cilfacade.get_ikind_exp exp in
+    if not (Cil.isSigned ik) then false else
+    if GobConfig.get_string "sem.int.signed_overflow" = "assume_none" then true else
+      let eval_int ctx exp =
+        match ctx.ask (Queries.EvalInt exp) with
+        | x when Queries.ID.is_int x -> Queries.ID.to_int x
+        | _ -> None
+      in
+      let int_min, int_max = IntDomain.Size.range ik in
+      match IntOps.BigIntOps.to_int64 int_min, IntOps.BigIntOps.to_int64 int_max with
+      | exception Failure _ -> false
+      | int_min, int_max ->
+        let min = eval_int ctx (BinOp (Lt, kinteger64 ik int_min, exp , intType)) in
+        let max = eval_int ctx (BinOp (Lt,  exp, kinteger64 ik int_max, intType)) in
+        match min, max with
+        | Some (x), _ when x = (IntOps.BigIntOps.of_int 1) -> true
+        | _, Some (x) when x = (IntOps.BigIntOps.of_int 1) -> true
+        | _, _ -> false
+
+  let no_overflow ctx exp =
+    let res = no_overflow ctx exp in
+    if M.tracing then M.tracel "no_ov" "no_ov %b exp: %s\n" res
+        (Pretty.sprint ~width:1 (Cil.printExp Cil.defaultCilPrinter () exp)); res
```

Comment:
Why do this here? The expression conversion already does overflow handling.

Reply:
We came up with an improved overflow handling that queries whether the other analyses might evaluate to top (eval_int and then bounds check) for expression assignments and if not, an overflow can definitely be excluded. This yields better precision for `affeq` because unlike polyhedra and octagon it can not determine any inequalities itself